packages/core/src/codewhisperer/service/transformByQ/transformApiHandler.ts

```diff
@@ -103,18 +103,41 @@ export async function uploadArtifactToS3(
     try {
         const uploadFileByteSize = (await nodefs.promises.stat(fileName)).size
         getLogger().info(
-            `Uploading project artifact at %s with checksum %s using uploadId: %s and size %s kB`,
+            `CodeTransformation: Uploading project artifact at %s with checksum %s using uploadId: %s and size %s kB`,
             fileName,
             sha256,
             resp.uploadId,
             Math.round(uploadFileByteSize / 1000)
         )
 
-        const response = await request.fetch('PUT', resp.uploadUrl, {
-            body: buffer,
-            headers: getHeadersObj(sha256, resp.kmsKeyArn),
-        }).response
-        getLogger().info(`CodeTransformation: Status from S3 Upload = ${response.status}`)
+        let response = undefined
+        const retriableCodes = [408, 429, 500, 502, 503, 504]
+        for (let i = 0; i < 3; i++) {
+            try {
+                response = await request.fetch('PUT', resp.uploadUrl, {
+                    body: buffer,
+                    headers: getHeadersObj(sha256, resp.kmsKeyArn),
+                }).response
+                getLogger().info(`CodeTransformation: upload to S3 status on attempt ${i + 1}/3 = ${response.status}`)
+                if (response.status === 200) {
+                    break
+                }
+                throw new Error('Upload failed')
+            } catch (e: any) {
+                if (response && !retriableCodes.includes(response.status)) {
+                    throw new Error(`Upload failed with status code = ${response.status}; did not automatically retry.`)
+                }
+                if (i !== 2) {
+                    await sleep(1000 * Math.pow(2, i))
+                }
+            }
+        }
+        if (!response || response.status !== 200) {
+            const uploadFailedError = `Upload failed after up to 2 retries with status code = ${response?.status ?? 'unavailable'}`
+            getLogger().error(`CodeTransformation: ${uploadFailedError}`)
+            throw new Error(uploadFailedError)
+        }
+        getLogger().info('CodeTransformation: Upload to S3 succeeded')
```

Comment:
I'm not a big fan of implementing our own retry logic rather than doing it at the SDK layer, but the existing S3 client's upload functionality requires a bucket name as a part of the request, and 1) I don't think we should have the internal S3 bucket names we use floating around this repo, 2) we use multiple different buckets, and 3) we don't have access to the bucket name in the response of our `CreateUploadUrl` — this API would have to be changed and then we can use the existing S3 client.

Reply:
Thanks for that context! That info is probably worth a code comment here or in the docstring of this function.

For reference, we have discussed some sort of util to "poll stuff", see https://taskei.amazon.dev/tasks/IDE-15280
[startDevEnvironmentWithProgress](https://github.com/aws/aws-toolkit-vscode/blob/98d8324ff95aa75e0d79c88e169414b66d024e86/packages/core/src/shared/clients/codecatalystClient.ts#L641) is a very complex case, maybe not exactly analogous to standard "request retry".